executor/load_stats.go

```diff
@@ -0,0 +1,127 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"encoding/json"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/util/chunk"
+	goctx "golang.org/x/net/context"
+)
+
+var _ Executor = &LoadStatsExec{}
+
+// LoadStatsExec represents a load statistic executor.
+type LoadStatsExec struct {
+	baseExecutor
+	info *LoadStatsInfo
+}
+
+// LoadStatsInfo saves the information of loading statistic operation.
+type LoadStatsInfo struct {
+	Path string
+	Ctx  context.Context
+}
+
+// loadStatsVarKeyType is a dummy type to avoid naming collision in context.
+type loadStatsVarKeyType int
+
+// String defines a Stringer function for debugging and pretty printing.
+func (k loadStatsVarKeyType) String() string {
+	return "load_stats_var"
+}
+
+// LoadStatsVarKey is a variable key for load statistic.
+const LoadStatsVarKey loadStatsVarKeyType = 0
+
+func (e *LoadStatsExec) exec(goCtx goctx.Context) error {
+	if len(e.info.Path) == 0 {
+		return errors.New("Load Stats: file path is empty")
+	}
+	val := e.ctx.Value(LoadStatsVarKey)
+	if val != nil {
+		e.ctx.SetValue(LoadStatsVarKey, nil)
+		return errors.New("Load Stats: previous load stats option isn't closed normally")
+	}
+	e.ctx.SetValue(LoadStatsVarKey, e.info)
+
+	return nil
+}
+
+// Next implements the Executor Next interface.
+func (e *LoadStatsExec) Next(goCtx goctx.Context) (Row, error) {
+	return nil, e.exec(goCtx)
+}
+
+// NextChunk implements the Executor NextChunk interface.
+func (e *LoadStatsExec) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	return errors.Trace(e.exec(goCtx))
+}
+
+// Close implements the Executor Close interface.
+func (e *LoadStatsExec) Close() error {
+	return nil
+}
+
+// Open implements the Executor Open interface.
+func (e *LoadStatsExec) Open(goCtx goctx.Context) error {
+	return nil
+}
+
+// Update updates the stats of the corresponding table according to the data.
+func (e *LoadStatsInfo) Update(data []byte) error {
+	jsonTbl := &statistics.JSONTable{}
+	if err := json.Unmarshal(data, jsonTbl); err != nil {
+		return errors.Trace(err)
+	}
+
+	do := domain.GetDomain(e.Ctx)
+	is := do.InfoSchema()
+
+	tableInfo, err := is.TableByName(model.NewCIStr(jsonTbl.DatabaseName), model.NewCIStr(jsonTbl.TableName))
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h := do.StatsHandle()
+	if h == nil {
+		return errors.New("Load Stats: handle is nil")
+	}
+
+	tbl, err := h.LoadStatsFromJSON(tableInfo.Meta(), jsonTbl)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, col := range tbl.Columns {
+		err = statistics.SaveStatsToStorage(e.Ctx, tbl.TableID, tbl.Count, 0, &col.Histogram, col.CMSketch)
```

Comment:
We should not directly call it because someone may write to the stats table at the same time.